# Step2 코어데이터 DB 구현 - jake, zziruru

jryoun1 opened the pull request:
@delmaSong 
안녕하세요~ 델마!! 🙋‍♀️🙋
금요일 밤에 PR을 날리게 되었습니당,,😭
코드 리뷰 부탁드립니다!! 

🐤 [고민사항]
1. CRUD를 진행할 때 마다(detailView→ListView로 갈때) **매번 core data를 fetch() + reloadTableView** 
VS **delegate pattern을 사용해서 바뀐 것만 수정**하는 방법 
- 이 부분에 대해서 저희는 메모의 양이 많은 경우에 매번 fetch()를 하게되면 오버헤드가 많을 것 같다고 생각했습니다.
그러나 실제로 iPhone, iPad에서 메모 앱을 사용해보면 뭔가 바뀐 부분만 변경하는 것보다는 
tableView가 변하는 animation이 fetch()하고 reload하는 것 같다는 느낌을 받았습니다.
일단 **구현은 delegate pattern을 사용해서 바뀐 부분만 수정하도록 구현**하였는데, 델마라면 어떠한 방법을 선택하실 것 같아요??

2. textView에서 **navigationItem의 `뒤로 가기 버튼`이나 keyboard toolbar의 `완료 버튼`을 눌렀을 때에 저장**이 되는 방법
VS textview의 text가 **매 번 수정이 될 때 마다 update()로 해서 저장**하는 방법
- 실제 iPhone, iPad 메모 앱에서는 매번 수정이 될 때마다 update()가 되는거 같아서 textViewDidChange()에서 update()를 하도록 구현
그러나 매 번 변경할 때마다 update()를 해주는 것이 overhead이지 않을까라는 고민을 해봐서, 이 부분에 대한 델마의 생각이 궁금합니다!!

3. DetailViewController가 굉장히 비대해진것 같은데, 역할을 분리해 볼 만한 지점이나, 코드를 따로 정리하는 방법이 있을지 궁금합니다.
- 저희는 일단 extension으로 유사한 작업들을 따로 분리하고 `//Mark: --` 를 사용해서 정리를 해보았습니다만 또 다른 방법이 있을까요??🤔

🐤 [구현사항]
1. CoreData Model 생성: `MemoModel`
   → 기존의 Memo, MemoData 모델 삭제하고 MemoModel 안에 list로 Memo를 가지게 됨

2. 더보기 버튼 선택 시 액션시트를 통해 ‘공유’ 및 ‘삭제’ 액션 표시
   → 공유 선택시 ‘액티비티 뷰’ 보여주도록 구현
   → 삭제 선택시 ‘Alert’ 띄워주도록 구현
   → iPad 일때는 actionsheet가 깨지므로 iPad와 iPhone 일때를 각각 다르게 구현

3. 메모 CRUD 구현
   → `Update`의 경우 즉각적으로 tableView에서 title, body의 내용이 수정되도록 구현
   → `Delete` 의 경우 tableView에서의 swipe action 시 삭제하는 경우도 포함

4. navigationBar 보이도록 수정
   → navigationBar를 만들어주기 위해서 showDetail로 넘길때 **navigationController를 하나 더 감싸서 보냄**
       (왜냐하면 showDetail은 masterViewController의 내용을 detailViewController로 replace하는 것이기 때문에 
       navigationBar가 없어지기 때문입니다.)

5. 저장되거나 업데이트 되는 시점
   → textViewDidChange() 호출 시 (내용이 변경될 때마다)

6. MemoUpdateDelegate 프로토콜 선언 및 구현
   → 선택한 memo 정보를 넘겨주는 memoSelected()
   → memo 삭제 시 textView를 비워주는 memoDeleted()

7. MemoListUpdateDelegate 프로토콜 선언 및 구현
   → save(), update(), delete() 를 각각 구현하여 tableview 전체를 reload하지 않고 수정, 삭제, 저장된 부분만 reload될 수 있도록 구현

jryoun1 committed:
Chore: 구현부 없는 메서드 삭제

jryoun1 committed:
Refactor: navigationBar가 보일수 있도록 수정

jryoun1 committed:
Feat: MasterViewController add버튼 눌렀을 때 emptyMemo를 보여줄 수 있도록 구현

jryoun1 committed:
Feat: NavigationBar에 더보기 버튼 추가 및 actionSheet와 alert창 구현

jryoun1 committed:
Refactor: IntExtension에서 DateExtension으로 변경

jryoun1 committed:
Feat: Memo Entity 생성 및 코드로 빼기

jryoun1 committed:
Refactor: Memo property를 modifiedDate → registerDate 로 수정

jryoun1 committed:
Refactor: protocol의 매개변수를 Memo 타입에서 Int 타입으로 변경

jryoun1 committed:
Refactor: keyboard frame 설정 및 textView에서 glyphIndex 범위 밖에 눌러도 인식할 수 있도록 수정

jryoun1 committed:
Chore: 필요없는 코드 삭제

jryoun1 committed:
Refactor: Memo, MemoData 모델 삭제하고 MemoModel 모델 생성 및 fetch() 함수 구현

jryoun1 committed:
Feat: MemoModel에서 save 메서드 구현 및 DetailViewController에서 저장할 수 있도록 구현

jryoun1 committed:
Feat: MemoModel에서 delete 메서드 구현 , Delegate pattern 사용해서 ListView의 view 변경할 수 있도록 구현

jryoun1 committed:
Feat: MemoModel에서 update 메서드 구현

jryoun1 committed:
Chore: 파일정리

jryoun1 committed:
Merge branch 'step2-jake-coredata' into steo2-jake

jryoun1 committed:
Refactor: 새로운 메모 생성시 Listview에 바로 업데이트 되도록 memoSelectetionDelegate의 매개변수를 Optional로 변경하여 구현

jae57 committed:
Refactor: memoSelected delegate method 수정

jae57 committed:
Refactor: alert 창에서의 삭제 수정

jae57 committed:
Chore: 코드 순서 변경

jae57 committed:
Refactor: update() 시점을 viewWillAppear()가 아닌 textViewDidChange()로 변경

jae57 committed:
Refactor: memoSelectionDelegate를 memoUpdateDelegate로 수정

jae57 committed:
Chore: 필요없는 주석 제거

jae57 committed:
Chore: ListViewController  중복코드 리팩토링

jae57 committed:
Chore: 메서드 위치 수정

jae57 committed:
Chore: refreshUI 메서드 분리

jae57 committed:
Chore: textViewDidChange() 분리

jae57 committed:
Chore: 폴더 및 파일 정리

delmaSong commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,261 @@
+    private func setDefaultMemo() {
+        MemoModel.shared.save(title: "새로운메모", body: "아직 내용없음")
+        memoBodyTextView.text = ""
+        delegate?.saveMemo(0)
```
새로운 메모를 생성하고 아무것도 입력하지 않았다면 아무것도 저장이 되지 않게끔 하는 건 어떨까요?

delmaSong commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,261 @@
+        content.append(NSAttributedString(string: "\n" + body, attributes: [NSAttributedString.Key.font: UIFont.preferredFont(forTextStyle: .body)]))
+        
+        return content
+    }
```
이부분을 좀 응용해보면 텍스트가 입력될 때도 타이틀과 바디를 구분해서 폰트 스타일을 적용할수도 있을 것 같아요 👀

jae57 replied:
폰트스타일을 적용하고 나면 뷰가 새로고침되면서 커서가 맨 뒤로 가는 문제가 있었는데 textViewDelegate 의 
```swift
func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool
```
를 이용하여 커서를 입력하던 부분으로 재조정해주도록 수정하였습니다.

delmaSong commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,261 @@
+            self?.shareMemo()
+        }
+        let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
+        let delete = UIAlertAction(title: "Delete", style: .destructive) { [weak self] _ in
```
요구사항 문서에 `Share`, `Cancel`, `Delete`로 나와있긴 한데 앱을 만들 때 제공하는 언어는 되도록 하나로 통일하는게 추후 로컬라이징을 이나 앱 전체의 일관성을 고려했을 때 더 좋을 것 같아요!

delmaSong commented on CloudNotes/CloudNotes/Controller/ListViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    private func setupDetailViewNavigationBar() {
+        if let detailViewController = delegate as? DetailViewController,
+           (traitCollection.horizontalSizeClass == .compact && traitCollection.userInterfaceIdiom == .phone) {
+            let detailViewNavigationController = UINavigationController(rootViewController: detailViewController)
+            splitViewController?.showDetailViewController(detailViewNavigationController, sender: nil)
+        }
```
`splitViewController?.showDetailViewController(detailViewController, sender: nil)`로
네비게이션컨트롤러를 새로 생성하지 않고 띄워줘도 네비게이션바가 잘 보이지 않나요?

jae57 replied:
@delmaSong 
글이 조금 깁니다 델마! ㅎㅎ

` iPhone, Portrait 모드`로 구동 시 셀을 클릭하면 화면이 넘어가는 것과 `아이패드, Landscape모드`로 구동 시 셀을 클릭하면 네비게이션 바가 상단에 보이는 것 두가지가 충돌하여 전자가 될때는 후자가 안되고, 후자가 될때는 전자가 안되는 문제가 있어서
우선은 showDetailView() 함수 내에서 각각의 경우를 분기쳐서 해결해보았습니다.
```swift
private func showDetailView() {
        guard let detailViewController = delegate as? DetailViewController else {
            return
        }
        
        if UITraitCollection.current.horizontalSizeClass == .compact && traitCollection.userInterfaceIdiom == .phone {
            splitViewController?.showDetailViewController(detailViewController, sender: nil)
        }
        else {
            guard let detailViewNavigationController = detailViewController.navigationController else {
                return
            }
            splitViewController?.showDetailViewController(detailViewNavigationController, sender: nil)
        }
    }
```
하지만, iPhone 11을 Portrait 모드로 구동하여 사용하다가 중간에 Landscape모드로 바꾸게 되면 네비게이션 바가 상단에서 사라집니다. (처음부터 가로로 킬 경우에는 네비게이션 바가 잘 나옵니다)

Landscape 모드로 바꿔서 셀을 클릭했을 때 showDetailView() 내부에서 detailViewController 의 navigationController 를 찍어보았더니 nil 로 나오는데 저희가 추측하기로는 splitViewController?.showDetailViewController(detailViewController, sender: nil) 이 함수자체가 splitViewController 의 detailViewController의 자리에 기존의 연결을 끊고 파라미터로 넘겨지는 새로운 detailViewController 를 붙이기 때문에 중간에 navigationController 를 잃어버린 것이라고 생각했습니다. 그래서 그 경우 다시 새로 NavigationController를 감싸서 넘겨주도록 해결해보았습니다. 저희가 추측한 것이 맞는지 델마의 의견이 궁금합니다!

[최종결론]
수정한 방법으로 했을 때 portrait 모드에서 셀을 선택했다가 landscape 모드로 돌리게되면 이미 네비게이션 컨트롤러가 없는 상태여서 새로 감싸주거나 하는 것이 어려워서 저희가 다시 결론 내린 것은 다시 원래처럼 navigation controller 를 하나 감싸주는 걸로 돌아왔는데 델마가 코멘트 남겨주셨던 이유가 어떤 것 때문이었는지 궁금하고, 이부분에 대해서 어떻게 생각하시는지가 궁금합니다!

글이 너무 길어서 이해가 안가는 부분이 있으시다면 언제든 DM 주시면 같이 이야기 해봐도 좋을 것 같습니다🙋‍♂️

delmaSong replied:
가로로 봤을 때 네비게이션바가 숨겨지는 건 미처 못보고 코멘트를 드렸네요🙇🏻‍♀️
왜 가로로 봤을 때는 네비게이션 바가 보여지지 않는지 고민해봤는데, 말씀 주신 것처럼 ` splitViewController?.showDetailViewController(_:sender:)` 이 함수 호출시 기존 연결된 `navigationController`의 연결을 끊어버린다기 보다는
- portrait일 때는 셀을 선택한 경우 보여지는 `DetailViewController` 상단의 네비게이션 바는 `SplitViewController`의 하위 뷰컨들을 지정해줄 때 설정된 (우리 코드의) `leftNavigationController`(== `ListViewController`의 상위 `UINavigationController`) 의 네비게이션 바이고
- landscape일 때 셀을 선택한 경우에는 `navigationController`로 감싸지지 않은 `detailViewController` 자체를 불러왔기 때문에 네비게이션 바가 보여지지 않은 것 같습니다. 

그래서 이부분은 `DetailViewController`의 상위뷰로 있는 `UINavigationController`를 띄워줌으로써 해결 할 수 있겠네요.
`ListViewController`나 `DetailViewController` 둘다 이미 네비게이션 컨트롤러의 하위 뷰컨이므로 새롭게 네비뷰컨을 생성해서 넣어주지는 않아도 될 것 같습니다
고민 많이하셨네요. 수고하셨습니다

delmaSong commented on CloudNotes/CloudNotes/Controller/ListViewController.swift:
```diff
@@ -0,0 +1,91 @@
+        
+        MemoModel.shared.delete(index: indexPath.row)
+        self.tableView.deleteRows(at: [indexPath], with: .fade)
+        self.delegate?.memoDeleted()
```
밀어서 삭제하기 좋네요 🤓👏🏼

delmaSong commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,261 @@
+        textView.dataDetectorTypes = [.link, .phoneNumber, .calendarEvent]
+    }
+    
+    private func splitLine(_ content: String) -> (String, String) {
+        let lines = content.split(separator: "\n", maxSplits: 1)
+        let title = String(lines[0])
+        var body = ""
+        
+        if lines.count > 1 {
+            body = String(lines[1])
+        }
+        
```
첫번 째 줄을 건너 뛰고 글을 작성하면 어떻게 될까요? 

크래시가 나는 부분을 최소화하기 위해선 배열의 맨 첫번째나 마지막에 있는 요소를 가져올 땐
`first`, `last`를 사용하는 것이 좋습니다.
혹은 아래처럼 subscript를 안전하게 사용하도록 확장해서 쓸 수도 있구요
```swift
subscript(safe index: Index) -> Iterator.Element? {
	guard indices.contains(index) else { return nil }
  return self[index]
}
```

delmaSong commented on CloudNotes/CloudNotes/AppDelegate.swift:
```diff
@@ -21,67 +21,35 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-                 * The store could not be migrated to the current model version.
-                 Check the error message to determine what the actual problem was.
-                 */
                 fatalError("Unresolved error \(error), \(error.userInfo)")
```
fatalError가 발생하면 앱이 크래시됩니다.
앱의 사용성을 고려해봤을 때, 에러가 발생하더라도 크래시가 나서 무작정 꺼지는 상황은 거의 절대로 없어야 한다고 생각해요.
그런 의미에서 크래시를 유발할 수 있는 fatalError를 코드에 내포하기보다는 혹여 앱을 껐다가 다시 키더라도 어떤식으로든 사용자가 다시 시도를 할 수 있게끔 유도해주는 상황을 만드는 것이 좋다고 생각합니다.
주석에도 에러가 발생할 수 있는 상황에 대한 설명들이 있으니까 fatalError를 내뱉는 방법 말고 다른 방법을 시도해 보면 어떨까요?

jae57 replied:
@delmaSong 

persistentContainer 가 생성되는 타이밍에 아직 SplitViewController가 load 만 되어있고 뷰가 올라온 상태가 아니라면 Alert창이 화면에 보여지지 않는 문제가 있었습니다.

이를 Notification 을 이용하여 viewDidAppear 에서 뷰가 hierarchy 에 올라갔음을 appDelegate에 알리고 이때 appDelegate 내부의 receiveNotification() 을 불러서 아래 코드에서 fatalError() 가 발생할 수 있는 시점에서 alert를 띄워서 사용자가 앱을 종료할 수 있도록 안내하였습니다.
```swift
container.loadPersistentStores(completionHandler: { (storeDescription, error) in                 
       if let error = error as NSError? {
       fatalError()    
}
```
[ 고민점 ]
1. 앱의 사용성 관련하여, alert창에 확인 버튼을 두어 사용자가 창을 종료할 수 있도록 하는 게 좋을지, 확인 버튼을 아예 빼서 더이상 동작을 진행시키지 않고 강제종료할 수 있도록 하는게 좋을지 고민이 됩니다! 현재는 확인 버튼을 빼두었습니다!

2. 만약 확인 버튼이 있다면 handler 에서 fatalError를 다시 발생시켜서 알아서 꺼지도록 하는게 좋을지 아니면 아무것도 넣지 말고 사용자가 메시지에 맞게 알아서 종료하도록 두는게 좋을지가 고민이 되었습니다

3. 그리고 이런 alert창을 발생시켜 fatalError 를 우회하는 방법 말고 델마가 주로 사용하는 방법은 어떤 것들이 있는지 궁금합니다!

delmaSong replied:
> 1. 앱의 사용성 관련하여, alert창에 확인 버튼을 두어 사용자가 창을 종료할 수 있도록 하는 게 좋을지, 확인 버튼을 아예 빼서 더이상 동작을 진행시키지 않고 강제종료할 수 있도록 하는게 좋을지 고민이 됩니다! 현재는 확인 버튼을 빼두었습니다!
> 2. 만약 확인 버튼이 있다면 handler 에서 fatalError를 다시 발생시켜서 알아서 꺼지도록 하는게 좋을지 아니면 아무것도 넣지 말고 사용자가 메시지에 맞게 알아서 종료하도록 두는게 좋을지가 고민이 되었습니다

예를들어 서버에 요청을 하고 응답값을 받아올 때에 이러한 에러가 발생했다면 확인 버튼을 누른 경우 재요청을 한다던가의 동작을 할 수 있을 것 같은데, 앱을 껐다가 다시 켜야만 하는 상황이라면 확인 버튼의 유무는 크게 중요하지 않을 것 같아요. 알럿을 띄움으로써 `무슨 문제가 있고 재시도가 필요한 상황이다` 를 사용자에게 인지시켜 준 걸로도 (무조건 개선해야하는 문제이지만!! 무작정 크래시나서 꺼지지만 않은걸로도) 어느정도의 사용성이 개선되었다고 볼 수 있을 것 같습니다.

>3. 그리고 이런 alert창을 발생시켜 fatalError 를 우회하는 방법 말고 델마가 주로 사용하는 방법은 어떤 것들이 있는지 궁금합니다!

제 경험상 앱을 잘 쓰고있다가 아무런 특별한 이유 없이 앱이 꺼져버리게 되면 그 앱을 잘 안쓰게되고 도대체 어떻게 만들었나(ㅋㅋ) 많이 답답하고 불친절한 느낌을 받았습니다. 그런데 알럿이나 토스트메시지를 띄워서 에러 발생하는 상황을 빠져나오거나 해결할 수 있게끔 우회하도록 도와준다거나 그마저도 안되어서 `잠시 후에 다시 시도해달라` 는 안내말이어도 그런 느낌은 좀 덜했어요.
그래서 저도 이러한 제 경험에 비추어 개발 할 때 가능한 크래시가 발생할만한 상황은 최대한 피하려고 하고, 에러가 발생할 수 있을만한 부분이라면 추후 트래킹을 위해 로그를 남기거나, 토스트나 알럿으로 사용자 인지 + 우회나 재시도 할 수 있게끔 해 사용성 ⬆️ 에러발생 ⬇️ 을 도모하고 있습니다..
그리고 덧붙여 크래시가 발생하지 않게끔 개발할때도 주의하는 것이 필요한데, 
1. 되도록 모든 코드에서 옵셔널 강제추출은 사용하지 않는다.
2. if let, guard let 작성하는 걸 귀찮아하지 않는다.
3.  배열에 접근할 때도 안전한 방법으로 추출한다.

등을 기억하시면서 개발하면 더 좋을 것 같습니다

delmaSong reviewed (commented):
두분 모두 한 주 내내 열심히 공부하고 구현하느라 수고 많으셨어요 😇
몇가지 개선점이나 생각해보면 좋을만한 점에 대해 코멘트 남겨놓았으니 확인해주세요!

`🐤 고민사항 `
> 1. CRUD를 진행할 때 마다(detailView→ListView로 갈때) 매번 core data를 fetch() + reloadTableView VS delegate pattern을 사용해서 바뀐 것만 수정하는 방법 

지금 구현 하신 것들을 보면 List는 Detail에 Detail은 List에 서로 서로 잡혀있는? 상태더라구요 
지금 `ListViewController`가 `MemoModel.shared.fetch()`로 메모 목록을 불러오는데, `DetailViewController`에서 어떤 변경이 일어난 경우 `delegate`로 `ListViewController`에 전달해주는 것 말고, 메모 목록쪽에서 변경사항이 일어난 경우에 `ListViewController`의 목록이 변경되도록 하는 건 어떨까요? 
List와 Detail간의 결합도를 낮추고 어차피 싱글톤으로 구현되어있으니 데이터의 변경점은 한곳만 바라보고 있게끔 하는게 어떨까 싶어요 

> 2. textView에서 navigationItem의 뒤로 가기 버튼이나 keyboard toolbar의 완료 버튼을 눌렀을 때에 저장이 되는 방법 VS textview의 text가 매 번 수정이 될 때 마다 update()로 해서 저장하는 방법

기본 메모 앱이 후자처럼 되어있어서 그런지 UX에 좀 더 익숙한 방법도 후자인 것 같아요 ㅎㅎ 저라도 제이크와 찌루루처럼 구현했을 것 같네요. 오버헤드를 많이 발생시키지 않을까 하는 걱정도 있지만 이 앱이 여러 사용자가 이용하고 이러한 변경사항에 대해 계속해서 서버로 요청을 보내야 하는 것이 아니라 한 시점에 하나의 변경사항만 로컬에서 업데이트 해주는 것이라서 유의미하게 오버헤드를 발생시키지는 않을 것 같다는 생각입니다!

> 3. DetailViewController가 굉장히 비대해진것 같은데, 역할을 분리해 볼 만한 지점이나, 코드를 따로 정리하는 방법이 있을지 궁금합니다.

저도 익스텐션으로 코드를 분리하는 방법을 종종 사용하고 그럴 땐 `// MARK: - SomeDelegate` 처럼 마크업 주석을 남기는 편입니다. 저렇게 대시(-)를 이용하면 quick navigation에서도 선이 그어지면서 구분을 해줘서 보기 편하다는 장점이 있어요 ㅎㅎ

jryoun1 committed:
Chore: 주석 MARK에 - 사용

jryoun1 committed:
Refactor: setupDetailViewNavigationBar() → showDetailView()로 변경

jryoun1 committed:
Refactor: showDetailView()에서 만약 navigationController가 없는 경우에는 생성해서 보내줄 수 있도록 수정

jryoun1 committed:
Refactor: showDetailView() 다시 원래 코드로 복구

jryoun1 committed:
Chore: ActionSheet의 title을 영어에서 한글로 수정

jryoun1 committed:
Refactor: Memo Entity 수정 및 MemoModel 변경 (title, body 두 개의 Attribute를 → content Attribute 하나로 변경)

jryoun1 committed:
Refactor: splitContent() 구현하여 setUpMemoCell에서 title, body 설정할 수 있도록 수정

jryoun1 committed:
Refactor: MemoModel 수정으로 인한 수정 및 applyFontStyle() 수정

jryoun1 committed:
Merge branch 'step2-jake-fix' into steo2-jake

jae57 committed:
Refactor: listview 와 detailview 간의 결합도 낮추기 위해 delegate -> notification으로 변경

jae57 committed:
Feat: 입력과 동시에 font style 적용되도록 수정

jryoun1 committed:
Feat: Core Data 관련 fatalError를 alert를 띄워서 앱을 종료할 수 있도록 구현

jryoun1 committed:
Refactor: 기존의 window가 2번 생성되었던 것을 발견하여 이를 수정

iOS 13.0이상의 버전에서는 Scene에서 생성이 되고, 그 이하의 버전에서는 appDelegate에서 window를 생성할 수 있도록 구현

jryoun1 committed:
Refactor: iOS13.0이후 버전과 그 이전 버전에 따라서 Alert창을 띄워줄 수 있도록 수정

jryoun1 committed:
Refactor: fatalError을 발생시키지 않고 Alert창을 띄워서 종료했다가 다시 실행할 수 있도록 수정

jae57 committed:
Refactor: content 가 nil 일 때 셀에 기본값 설정하도록 수정

jae57 committed:
Feat: add 버튼 토글기능 추가

jae57 committed:
Feat: 메모에 내용이 없을 경우 자동 삭제

jae57 committed:
Refactor: applyFontStyle() 메서드 분리 및 네이밍 수정

jryoun1 committed:
Merge branch 'steo2-jake' of https://github.com/jryoun1/ios-cloud-notes into steo2-jake

jryoun1 commented:
@delmaSong 
델마! 안녕하세요~ 
다시 한번 코드 리뷰 부탁드려요!! 

**🐤    [고민사항]**
> 지금 구현 하신 것들을 보면 List는 Detail에 Detail은 List에 서로 서로 잡혀있는? 상태더라구요
지금 ListViewController가 MemoModel.shared.fetch()로 메모 목록을 불러오는데, DetailViewController에서 어떤 변경이 일어난 경우 delegate로 ListViewController에 전달해주는 것 말고, 메모 목록쪽에서 변경사항이 일어난 경우에 ListViewController의 목록이 변경되도록 하는 건 어떨까요?
List와 Detail간의 결합도를 낮추고 어차피 싱글톤으로 구현되어있으니 데이터의 변경점은 한곳만 바라보고 있게끔 하는게 어떨까 싶어요

델마의 리뷰를 보고 나니까 실제로 서로가 서로를 알게 되면 결합도가 높아지는 것 같아서, 아래와 같은 방향으로 수정하였습니다.
기존의 `Detail`에서 `List`로 **Delegate 패턴**을 사용했던 부분을 
`MemoModel`에 **Notification을 활용**하여 변화가 있으면 List에서 View를 변경할 수 있도록 수정하였습니다!!

하지만 여기서 고민해 볼만한 부분이 생겼는데, 각각의 CRUD에 대해서 Notification을 활용하다보니까 
너무 많은 Notification을 사용하는 것이 아닌가에 대한 의문이 생겼습니다. 
그래서 이렇게 Notification을 여러 개 사용하였을 때의 **성능적으로의 문제**라던가, 아니면 **구조적으로의 문제**가 있을지에 대해서 궁금합니다!!

그리고 Notification을 대체할 수 있는 방법이 있을지에 대해서도 좋은 의견이 있다면 알려주시면 정말 감사하겠습니다!!🙇‍♂️🙇‍♀️

그리고 윗 부분에 코멘트로도 고민사항을 몇 개 남겨놓았습니다!!🙋🙋‍♀️

**🐤    [구현사항]** (자세한 사항들은 코멘트 댓글로 달아 두었습니다:wink:)
- 새로운 메모를 생성하고 아무것도 입력하지 않았다면 아무것도 저장이 되지 않도록 수정 
- ListView가 DetailView가 아닌 데이터 모델에서 noti를 받아 뷰를 업데이트 하도록 수정
- 텍스트가 입력될 때도 타이틀과 바디를 구분해서 폰트 스타일을 적용할수있도록 수정
  → UITextViewDelegate 내 함수를 override 하여 커서 위치가 그대로 유지되도록 구현
- 언어 한국어 하나로 통일
- splitLine() 내에서 인덱스를 바로 접근해서 크래시 날 수 있는 부분을 first 이용하도록 수정
- appDelegate 파일 내의 fatalError() 를 alert 창을 띄우도록 수정

delmaSong commented on CloudNotes/CloudNotes/AppDelegate.swift:
```diff
@@ -13,75 +13,45 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-                 Check the error message to determine what the actual problem was.
-                 */
-                fatalError("Unresolved error \(error), \(error.userInfo)")
+                NotificationCenter.default.addObserver(self, selector: #selector(self.receiveNotification(_:)), name: Notification.Name("loadPersistentStoresError"), object: nil)
```
이 부분이 에러를 발생시키는 쪽 아닌가요? 관찰하고 있네요

delmaSong commented on CloudNotes/CloudNotes/AppDelegate.swift:
```diff
@@ -13,75 +13,45 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     }
-
+    
+    @objc func receiveNotification(_ notification: Notification) {
+        let alert = UIAlertController(title: "에러가 발생했습니다.", message: "앱을 종료했다가 다시 켜주세요!", preferredStyle: .alert)
+        
```
그리고 알럿을 띄워주는 부분이라 여기보다는 뷰쪽이 더 적당하지 않을까 싶어요 🤔

delmaSong commented on CloudNotes/CloudNotes/Controller/ListViewController.swift:
```diff
@@ -0,0 +1,137 @@
+        navigationItem.rightBarButtonItem?.isEnabled = false
+    }
+}
+
+extension Notification.Name {
+    static let createMemo = Notification.Name("createMemo")
+    static let deleteMemo = Notification.Name("deleteMemo")
```
익스텐션 활용 좋네요 👍🏼

delmaSong commented on CloudNotes/CloudNotes/Controller/SplitViewController.swift:
```diff
@@ -14,6 +14,11 @@ final class SplitViewController: UISplitViewController, UISplitViewControllerDel
         setUpViewControllers()
     }
     
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        NotificationCenter.default.post(name: NSNotification.Name("loadPersistentStoresError"), object: nil)
+    }
```
제 생각엔 이쪽이 해당 노티를 관찰하는 게 더 적합할 것 같아요

delmaSong reviewed (commented):
제이크, 찌루루 안녕하세요! 수고 많으셨습니다.
Delegate 패턴 부분을 Notification을 이용하는 걸로 바꾸셨군요! 
Delegate와 Notification은 각각 어떤 상황에서 사용하면 좋을까요?

 만약 지금같은 방법에 뷰가 여러개 더 늘어난다고 하면 개개의 뷰끼리 데이터를 들고 소통을 해야하는데 그럼 그로인해 따라붙어야 하는 코드가 너무 많아지고 관리하기 어렵지 않을까 싶은데 어떻게 생각하시나요?
 지난번 드렸던 코멘트의 `데이터의 변경점이 한곳만 바라보고 있다` 는 말에 설명을 좀 덧붙이자면, 아래의 그림처럼 `ListViewController`와 `DetailViewController`, `MemoModel`이 있을때 데이터의 변경사항은 전부 `MemoModel`을 거치게 됩니다. 그러므로 `ListVC`와 `DetailVC`가 데이터를 들고 직접적으로 소통하기보단 `ListVC`와 `DetailVC`가 `MemoModel`을 바라보고 있고 변경점이 발생한 경우에 `ListVC`를 업데이트 해주는 방향이 어떨까 싶어서 드린 말씀이었습니다. 예를 들어 `DetailVC`에서 `MemoModel`의 `delete(index:)`를 호출하면 `MemoModel`의 `list`에 변화가 생길텐데 그럼 그때 `ListViewController`가 자신의 테이블뷰를 업데이트하는 식이 될 수 있을 것 같군요.
그렇게 하면 `ListVC`와 `DetailVC`는 화면을 띄우고 사라지게 할때의 최소한의 데이터만 건네받도록 만들 수 있겠네요.

<img width="60%" src="https://user-images.githubusercontent.com/40784518/109836017-19c98b00-7c87-11eb-9929-2c6654c22f82.jpeg"/>

delmaSong commented:
앞선 코멘트와 회의실에서 의논했던 부분은 스텝3에서 적용해주셔요~! 수고하셨습니다 🙇🏻‍♀️
